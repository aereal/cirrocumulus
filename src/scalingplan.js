"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const cdk_1 = require("@aws-cdk/cdk");
const aws_autoscalingplans_1 = require("@aws-cdk/aws-autoscalingplans");
class ScalingPlan extends cdk_1.Construct {
    constructor(scope, id, props) {
        super(scope, id);
        const minCapacity = props.minCapacity ? props.minCapacity : 1;
        const maxCapacity = props.maxCapacity ? props.maxCapacity : minCapacity;
        new aws_autoscalingplans_1.CfnScalingPlan(scope, 'AutoScalingPlan', {
            applicationSource: {
                tagFilters: props.tagFilters
            },
            scalingInstructions: [{
                    minCapacity,
                    maxCapacity,
                    resourceId: `autoScalingGroup/${props.autoScalingGroupName}`,
                    scalableDimension: "autoscaling:autoScalingGroup:DesiredCapacity",
                    scalingPolicyUpdateBehavior: "ReplaceExternalPolicies",
                    serviceNamespace: "autoscaling",
                    targetTrackingConfigurations: [{
                            predefinedScalingMetricSpecification: {
                                predefinedScalingMetricType: "ASGAverageCPUUtilization"
                            },
                            estimatedInstanceWarmup: 300,
                            targetValue: props.targetPercentage ? props.targetPercentage : 50
                        }]
                }]
        });
    }
}
exports.ScalingPlan = ScalingPlan;
