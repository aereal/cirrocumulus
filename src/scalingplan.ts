import { Construct } from "@aws-cdk/cdk"

import { CfnScalingPlan } from "@aws-cdk/aws-autoscalingplans"

export interface ScalingPlanProps {
  /**
   * Rhe name of the Auto Scaling group
   *
   */
  autoScalingGroupName: string

  /**
   * A set of tags
   *
   */
  tagFilters: CfnScalingPlan.TagFilterProperty[]

  /**
   * The minimum capacity of the resource
   *
   * @default 1
   */
  minCapacity?: number

  /**
   * The maximum capacity of the resource
   *
   * @default 1
   */
  maxCapacity?: number

  /**
   * The target value for the metric
   *
   * @default 50
   */
  targetPercentage?: number
}

export class ScalingPlan extends Construct {
  constructor(scope: Construct, id: string, props: ScalingPlanProps) {
    super(scope, id)

    const minCapacity = props.minCapacity ? props.minCapacity! : 1
    const maxCapacity = props.maxCapacity ? props.maxCapacity! : minCapacity

    const scalingPlan = new CfnScalingPlan(scope, "Resource", {
      applicationSource: {
        tagFilters: props.tagFilters,
      },
      scalingInstructions: [
        {
          maxCapacity,
          minCapacity,
          resourceId: `autoScalingGroup/${props.autoScalingGroupName}`,
          scalableDimension: "autoscaling:autoScalingGroup:DesiredCapacity",
          scalingPolicyUpdateBehavior: "ReplaceExternalPolicies",
          serviceNamespace: "autoscaling",
          targetTrackingConfigurations: [
            {
              estimatedInstanceWarmup: 300,
              predefinedScalingMetricSpecification: {
                predefinedScalingMetricType: "ASGAverageCPUUtilization",
              },
              targetValue: props.targetPercentage ? props.targetPercentage : 50,
            },
          ],
        },
      ],
    })
  }
}
