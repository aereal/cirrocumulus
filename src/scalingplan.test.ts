import { SynthUtils } from "@aws-cdk/assert"
import { Stack } from "@aws-cdk/cdk"
import { ScalingPlan } from "./scalingplan"

describe("scalingplan", () => {
  test("default", () => {
    const stack = new Stack()

    new ScalingPlan(stack, "ScalingPlan", {
      autoScalingGroupName: "my-asg",
      tagFilters: [{
        key: "aws:cloudformation:stack-name",
        values: [stack.name]
      }]
    })

    expect(SynthUtils.toCloudFormation(stack)).toMatchSnapshot()
  })

  test("can set capacities", () => {
    const stack = new Stack()

    new ScalingPlan(stack, "ScalingPlan", {
      autoScalingGroupName: "my-asg",
      minCapacity: 2,
      maxCapacity: 4,
      tagFilters: [{
        key: "aws:cloudformation:stack-name",
        values: [stack.name]
      }]
    })

    expect(SynthUtils.toCloudFormation(stack)).toMatchSnapshot()
  })

  test("can set a targetPercentage", () => {
    const stack = new Stack()

    new ScalingPlan(stack, "ScalingPlan", {
      autoScalingGroupName: "my-asg",
      targetPercentage: 60,
      tagFilters: [{
        key: "aws:cloudformation:stack-name",
        values: [stack.name]
      }]
    })

    expect(SynthUtils.toCloudFormation(stack)).toMatchSnapshot()
  })
})
