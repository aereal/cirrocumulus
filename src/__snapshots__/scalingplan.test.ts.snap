// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`scalingplan can set a targetPercentage 1`] = `
Object {
  "Resources": Object {
    "AutoScalingPlan": Object {
      "Properties": Object {
        "ApplicationSource": Object {
          "TagFilters": Array [
            Object {
              "Key": "aws:cloudformation:stack-name",
              "Values": Array [
                "Stack",
              ],
            },
          ],
        },
        "ScalingInstructions": Array [
          Object {
            "MaxCapacity": 1,
            "MinCapacity": 1,
            "ResourceId": "autoScalingGroup/my-asg",
            "ScalableDimension": "autoscaling:autoScalingGroup:DesiredCapacity",
            "ScalingPolicyUpdateBehavior": "ReplaceExternalPolicies",
            "ServiceNamespace": "autoscaling",
            "TargetTrackingConfigurations": Array [
              Object {
                "EstimatedInstanceWarmup": 300,
                "PredefinedScalingMetricSpecification": Object {
                  "PredefinedScalingMetricType": "ASGAverageCPUUtilization",
                },
                "TargetValue": 60,
              },
            ],
          },
        ],
      },
      "Type": "AWS::AutoScalingPlans::ScalingPlan",
    },
  },
}
`;

exports[`scalingplan can set capacities 1`] = `
Object {
  "Resources": Object {
    "AutoScalingPlan": Object {
      "Properties": Object {
        "ApplicationSource": Object {
          "TagFilters": Array [
            Object {
              "Key": "aws:cloudformation:stack-name",
              "Values": Array [
                "Stack",
              ],
            },
          ],
        },
        "ScalingInstructions": Array [
          Object {
            "MaxCapacity": 4,
            "MinCapacity": 2,
            "ResourceId": "autoScalingGroup/my-asg",
            "ScalableDimension": "autoscaling:autoScalingGroup:DesiredCapacity",
            "ScalingPolicyUpdateBehavior": "ReplaceExternalPolicies",
            "ServiceNamespace": "autoscaling",
            "TargetTrackingConfigurations": Array [
              Object {
                "EstimatedInstanceWarmup": 300,
                "PredefinedScalingMetricSpecification": Object {
                  "PredefinedScalingMetricType": "ASGAverageCPUUtilization",
                },
                "TargetValue": 50,
              },
            ],
          },
        ],
      },
      "Type": "AWS::AutoScalingPlans::ScalingPlan",
    },
  },
}
`;

exports[`scalingplan default 1`] = `
Object {
  "Resources": Object {
    "AutoScalingPlan": Object {
      "Properties": Object {
        "ApplicationSource": Object {
          "TagFilters": Array [
            Object {
              "Key": "aws:cloudformation:stack-name",
              "Values": Array [
                "Stack",
              ],
            },
          ],
        },
        "ScalingInstructions": Array [
          Object {
            "MaxCapacity": 1,
            "MinCapacity": 1,
            "ResourceId": "autoScalingGroup/my-asg",
            "ScalableDimension": "autoscaling:autoScalingGroup:DesiredCapacity",
            "ScalingPolicyUpdateBehavior": "ReplaceExternalPolicies",
            "ServiceNamespace": "autoscaling",
            "TargetTrackingConfigurations": Array [
              Object {
                "EstimatedInstanceWarmup": 300,
                "PredefinedScalingMetricSpecification": Object {
                  "PredefinedScalingMetricType": "ASGAverageCPUUtilization",
                },
                "TargetValue": 50,
              },
            ],
          },
        ],
      },
      "Type": "AWS::AutoScalingPlans::ScalingPlan",
    },
  },
}
`;
